Report/Chapters/Design.tex

```diff
@@ -8,5 +8,50 @@ \section{Languages and libraries}
 
 \section{High-level architecture}
 
-% Frontend/backend split
-% Communication between frontend and backend (when we've got that settled)
+Viskell consists of two parts, a front end and a back end.
+The front end is responsible for providing a user interface, handling user input, displaying error messages, and communicating with the back end.
+The back end consists of a representation for Haskell code in Java, an interface to GHCi and functionality to make working with the back end easier.
+This separation is made to make development and debugging for the part that generates Haskell code easier.
+It also makes using the user interface for a different purpose a bit easier, although the user interface is not specifically designed for this purpose.
+
+\section{Backend architecture}
+
+A result of our design choices is that our Java program needs to have a basic understanding of the Haskell programming
+language. For this purpose we implemented two tree structures - one for expressions, one for types - supported by an
+interface for GHCi, a type signature parser and a catalog.
+
+\subsection{Communication with GHCi}
+\index{GHCi}
+
+The base of the back-end is an interface for communication with GHCi. This interface allows us to run Haskell code
+generated by the users of our application. As implementing a complete type checker for Haskell is a major challenge this
+interface can also be used to check a program for faults.
+
+\subsection{Type system and type checker}
+
+The error messages from GHCi are hard to parse and do not always provide detailed information about where the error
+occured. Therefore we implemented our own type checker. This type checker is designed to catch most of the type errors
+which are in the design scope of the project and not raise false negatives. Pushing Haskell code to GHCi is still needed
+to be sure that the code compiles (and will always be needed to catch runtime errors).
+
+Internally Haskell types are represented as instances of a class. These instances can be tied together to form a
+tree-like structure. This approach is chosen because it is easy to work with class instances and nested types form
+a tree-like structure. \index{type tree}
+
+The type checker is implemented using Hindley-Milner type inference. We use this algorithm to (try to) unify two
+types. If the types cannot be unified, there exists a type error. The algorithm is described in detail by
+\cite{borisov}. \index{Hindley-Milner} \index{type inference}
+
+\subsection{Expressions}
+
+Haskell expressions are represented in a tree. By design each function can have only one argument. Functions with more
+arguments are represented as a function with a single argument that produces another function with the other arguments.
```

Comment:
that produces another function with another argument and so on.

(Currently it is as if the second function does accept more than 1 argument)